### DISCUSSION TITLE
Local development with Android emulator, unable to connect to Auth service

### DISCUSSION THREAD

**Discussion by calebcng** [supabase/supabase#21174] [Questions]
Hi, I am newly exploring Supabase, coming from Firebase, and am just in the process of trying to setup my local development environment. 

At the moment I have a basic Android app with a Login screen written in Dart (Flutter), which uses Google OAuth to create a new user account. I've tested connecting to the Supabase platform and everything works fine on my Android emulator. However, when I try switching to local, I keep getting the following error:
`ClientException with SocketException: Connection refused (OS Error: Connection refused, errno = 111), address = 127.0.0.1, port = 46232, uri=http://127.0.0.1:54321/auth/v1/token?grant_type=id_token`
![image](https://github.com/supabase/supabase/assets/5797919/f84ec944-8c58-4369-8373-f94c639601d6)

I've looked over my `config.toml` and my Google Cloud settings and it seems correct as far as I can tell. Could anyone advise where else I should be looking?

My `config.toml` file is as follows:
```
...

[auth]
enabled = true
# The base URL of your website. Used as an allow-list for redirects and for constructing URLs used
# in emails.
site_url = "http://127.0.0.1:3000"
# A list of *exact* URLs that auth providers are permitted to redirect to post authentication.
additional_redirect_urls = ["https://127.0.0.1:3000","http://localhost:54321/auth/v1/callback"]
# How long tokens are valid for, in seconds. Defaults to 3600 (1 hour), maximum 604,800 (1 week).
jwt_expiry = 3600
# If disabled, the refresh token will never expire.
enable_refresh_token_rotation = true
# Allows refresh tokens to be reused after expiry, up to the specified interval in seconds.
# Requires enable_refresh_token_rotation = true.
refresh_token_reuse_interval = 10
# Allow/disallow new user signups to your project.
enable_signup = true
# Allow/disallow testing manual linking of accounts
enable_manual_linking = false

...

# Use an external OAuth provider. The full list of providers are: `apple`, `azure`, `bitbucket`,
# `discord`, `facebook`, `github`, `gitlab`, `google`, `keycloak`, `linkedin_oidc`, `notion`, `twitch`,
# `twitter`, `slack`, `spotify`, `workos`, `zoom`.
[auth.external.apple]
enabled = false
client_id = ""
# DO NOT commit your OAuth provider secret to git. Use environment variable substitution instead:
secret = "env(SUPABASE_AUTH_EXTERNAL_APPLE_SECRET)"
# Overrides the default auth redirectUrl.
redirect_uri = ""
# Overrides the default auth provider URL. Used to support self-hosted gitlab, single-tenant Azure,
# or any other third-party OIDC providers.
url = ""

[auth.external.google]
enabled = true
client_id = "<REDACTED>"
secret = "<REDACTED>"
# Overrides the default auth redirectUrl.
redirect_uri = "http://localhost:54321/auth/v1/callback"
# Overrides the default auth provider URL. Used to support self-hosted gitlab, single-tenant Azure,
# or any other third-party OIDC providers.
url = ""


[analytics]
enabled = true
port = 54327
vector_port = 54328
# Configure one of the supported backends: `postgres`, `bigquery`.
backend = "postgres"

...

```

**Comment by Ben-Vollrath** [comment]
Hi @calebcng I think your problem might be due to using localhost (127.0.0.1) as the supabase url on your emulator. This will send request to the emulators local host, and not to your dev machine where you are running your supabase local instance. 

Typically you would use 10.0.2.2 as the Host-Loopback for the dev machines localhost.
You can read up more here: https://developer.android.com/studio/run/emulator-networking

**Comment by Jonny1987** [comment]
I tried using 10.0.2.2 but now instead of a SocketException I just get a timeout error when signing in through my app:

```
Error: ClientException with SocketException: Connection timed out (OS Error: Connection timed out, errno = 110), address = 10.0.2.2, port = 36592, uri=http://10.0.2.2:54321/auth/v1/token?grant_type=password
```

**Reply by jan-tennert** [reply]
Can't reproduce. Just created a emulator within IntelliJ, started the Supabase instance via the CLI and started the supabase-kt/chat-demo sample and everything works. Shouldn't be different for Flutter :thinking: 
- Which OS are you on?
- How did you create the emulator?

**Reply by grdsdev** [reply]
Hi @jan-tennert I think the issue occurs when trying to use Google Sign-in locally, did you try that?

**Reply by Jonny1987** [reply]
Just added a comment how I got it worked. Fyi I had this problem when using a real device. Not sure if the problem exists when using an emulator or not

**Reply by jan-tennert** [reply]
Yea makes sense that it fails on a real device, shouldn't be an issue on an emulator.

**Comment by Jonny1987** [comment]
I got this working by keeping the supabase url as the default (127.0.0.1) running `adb reverse tcp:54321 tcp:54321` in the terminal
